evm/src/03_nft_burn_bridging/NftBurnBridging.sol

```diff
@@ -7,80 +7,201 @@ import "modules/utils/BytesLib.sol";
+		if (tokenCount < 2 || tokenCount > _maxBatchSize) {
+			revert InvalidBatchCount();
+		}
+
+		_burnAndSend(tokenIds, tokenCount, recipient);
+	}
+
+	function receiveAndMint(bytes calldata vaa) external {
+		IWormhole.VM memory vm = verifyMintMessage(vaa);
+
+		(uint256 tokenId, address evmRecipient) = parsePayload(vm.payload);
+		_safeMint(evmRecipient, tokenId);
+	}
+
+	function receiveAndMintBatch(bytes calldata vaa) external {
+		IWormhole.VM memory vm = verifyMintMessage(vaa);
+
+		(uint256[] memory tokenIds, address evmRecipient) = parseBatchPayload(vm.payload);
+
+		uint256 tokenCount = tokenIds.length;
+		for (uint256 i = 0; i < tokenCount; ) {
+			_safeMint(evmRecipient, tokenIds[i]);
+
+			unchecked {
+				i += 1;
+			}
+		}
+	}
+
+	function _burnAndSend(
+		uint256[] memory tokenIds,
+		uint256 tokenCount,
+		address recipient
+	) internal {
+		if (recipient == address(0)) {
+			revert RecipientZeroAddress();
+		}
+
+		uint256 lastTokenId;
+		bytes memory payload;
+		for (uint256 i = 0; i < tokenCount; ) {
+			uint256 tokenId = tokenIds[i];
+
+			//tokenIds must be ascending and unique
+			if (i != 0 && tokenId <= lastTokenId) {
+				revert NotAscendingOrDuplicated();
+			}
+
+			if (!_isApprovedOrOwner(_msgSender(), tokenId)) {
+				revert BurnNotApproved();
+			}
+
+			_burn(tokenId);
+
+			//add tokenId to the message payload
+			payload = abi.encodePacked(payload, uint16(tokenId));
+
+			unchecked {
+				lastTokenId = tokenId;
+				i += 1;
+			}
+		}
+
+		//append the recipient to the payload and send the message
+		_wormhole.publishMessage{value: msg.value}(
+			0, //nonce
+			abi.encodePacked(payload, recipient),
+			_finality
+		);
+	}
+
+	function verifyMintMessage(bytes calldata vaa) internal returns (IWormhole.VM memory) {
+		(IWormhole.VM memory vm, bool valid, string memory reason) = _wormhole.parseAndVerifyVM(
+			vaa
+		);
+		if (!valid) revert FailedVaaParseAndVerification(reason);
+
+		if (vm.emitterChainId != _emitterChainId) revert WrongEmitterChainId();
+
+		if (vm.emitterAddress != _emitterAddress) revert WrongEmitterAddress();
+
+		if (_claimedVaas[vm.hash]) revert VaaAlreadyClaimed();
+
+		_claimedVaas[vm.hash] = true;
+
+		return vm;
+	}
+
+	function parsePayload(
+		bytes memory message
+	) internal pure returns (uint256 tokenId, address evmRecipient) {
+		if (message.length != BytesLib.uint16Size + BytesLib.addressSize)
+		revert InvalidMessageLength();
+
+		tokenId = message.toUint16(0);
+		evmRecipient = message.toAddress(BytesLib.uint16Size);
+	}
+
+	function parseBatchPayload(
+		bytes memory message
+	) internal pure returns (uint256[] memory, address) {
+		uint256 messageLength = message.length;
+		uint256 endTokenIndex = messageLength - BytesLib.addressSize;
+		uint256 batchSize = endTokenIndex / BytesLib.uint16Size;
+
+		if (
+			messageLength <= BytesLib.uint16Size + BytesLib.addressSize ||
+			endTokenIndex % BytesLib.uint16Size != 0
+		) {
+			revert InvalidMessageLength();
+		}
```

Comment:
We don't check here or in the caller if the size > max batch, that fine for incoming?

Reply:
That check is done on the sending side in `_burnAndSend`. This allows two sending contracts (e.g. A and B) to both send batches of different length to contract C.